### PR TITLE
[DISCO-3005] Pass Language to weather report logic

### PR DESCRIPTION
**Pull request by tiftran**
## References

JIRA: [DISCO-3005](https://mozilla-hub.atlassian.net/browse/DISCO-3005)

## Description
Passed along the language header to weather report logic to be able to retrieve, current conditions and forecast in different languages.

## PR Review Checklist

_Put an `x` in the boxes that apply_

- [ ] This PR conforms to the [Contribution Guidelines](https://github.com/mozilla-services/merino-py/blob/main/CONTRIBUTING.md)
- [ ] The PR title starts with the JIRA issue reference, format `[DISCO-####]`, and has the same title (if applicable)
- [ ] `[load test: (abort|warn)]` keywords are applied (if applicable)
- [ ] [Documentation](https://github.com/mozilla-services/merino-py/tree/main/docs) has been updated (if applicable)
- [ ] [Functional and performance test](https://github.com/mozilla-services/merino-py/blob/main/docs/dev/testing.md) coverage has been expanded and maintained (if applicable)

[DISCO-3005]: https://mozilla-hub.atlassian.net/browse/DISCO-3005?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: merino/providers/weather/backends/accuweather/pathfinder.py [hunk 1]
```diff
@@ -47,7 +47,9 @@ def compass(location: Location) -> Generator[Triplet, None, None]:
 
 
 async def explore(
-    location: Location, probe: Callable[[MaybeStr, MaybeStr, MaybeStr], Awaitable[Optional[Any]]]
+    location: Location,
+    probe: Callable[..., Awaitable[Optional[Any]]],
+    language: str | None = None,
```
Comment (tiftran) [inline merino/providers/weather/backends/accuweather/pathfinder.py:52]: im not too thrilled with this... feels like cheating.

**Comment by tiftran** [comment]
I integrated `weather_context` into this PR, rerequesting review to see what people think

File: merino/providers/weather/backends/protocol.py [hunk 2]
```diff
@@ -72,6 +72,14 @@ class LocationCompletion(BaseModel):
...
 
 
+class WeatherContext:
+    """Class that contains context needed to make weather reports."""
+
+    def __init__(self, geolocation: Location, languages: list[str]):
+        self.geolocation = geolocation
+        self.languages = languages
+
+
```
Comment (Herraj) [inline merino/providers/weather/backends/protocol.py:82]: Nice! yea this feels concise! Also, no need to change just sharing but I also learned about this: [@dataclass](https://docs.python.org/3.12/library/dataclasses.html)

**Review by Herraj — APPROVED** [review]

File: merino/providers/weather/backends/protocol.py (hunk 2 above)
Comment (tiftran) [reply merino/providers/weather/backends/protocol.py:82]: oooh yes, i see we already use it inside merino, adding that.

**Review by gruberb — APPROVED** [review]

File: merino/providers/weather/backends/accuweather/backend.py [hunk 3]
```diff
@@ -284,9 +285,12 @@ def cache_key_for_accuweather_request(
         return f"{self.__class__.__name__}:v4:{url}"
 
     @functools.cache
-    def cache_key_template(self, dt: WeatherDataType) -> str:
+    def cache_key_template(self, dt: WeatherDataType, language: str) -> str:
```
Comment (ncloudioj) [inline merino/providers/weather/backends/accuweather/backend.py:288]: Let's bump the cache key version in `cache_key_for_accuweather_request()`.

Comment (tiftran) [reply merino/providers/weather/backends/accuweather/backend.py:288]: yup yup, remembered this morning 😅

**Review by ncloudioj — APPROVED** [review]
Great stuff, thanks!

**Pull request merged**